server/chat-plugins/smogtours.ts

```diff
@@ -0,0 +1,447 @@
+		tours: [],
+		desc: "Tournaments run by Smogon staff.",
+	},
+	smogon: {
+		title: "Open Sign-Ups",
+		tours: [],
+		desc: "Tournaments run by Smogon staff and regular users alike."
+	},
+	ps: {
+		title: "Pokémon Showdown!",
+		icon: ['https://play.pokemonshowdown.com/pokemonshowdownbeta.png', 146, 44],
+		tours: [],
+		desc: "Tournaments run by the rooms of Pokemon Showdown.",
+	},
+};
+try {
+	const data = JSON.parse(FS('config/chat-plugins/smogtours.json').readSync());
+	// settings should prioritize hardcoded values for these keys
+	const PRIO = ['title', 'icon'];
+	for (const key in data) {
+		const section = (tours[key] ||= data[key]) as any;
+		for (const k in data[key]) {
+			if (PRIO.includes(k)) {
+				if (!section[k]) section[k] = data[key][k];
+			} else {
+				section[k] = data[key][k];
+			}
+		}
+	}
+} catch {}
+
+function saveTours() {
+	FS('config/chat-plugins/smogtours.json').writeUpdate(() => JSON.stringify(tours));
+}
+
+function getTour(categoryID: ID, id: string) {
+	id = toID(id);
+	return tours?.[categoryID].tours.find(f => f.id === id) || null;
+}
+
+function checkWhitelisted(category: ID, user: User) {
+	return category ?
+		tours[category].whitelist?.includes(user.id) :
+		Object.values(tours).some(f => f.whitelist?.includes(user.id));
+}
+
+function checkCanEdit(user: User, context: Chat.PageContext | Chat.CommandContext, category?: ID) {
+	category = toID(category);
+	if (!checkWhitelisted(category, user)) {
+		context.checkCan('globalban');
+	}
+}
+
+export const commands: Chat.ChatCommands = {
+	smogtours: {
+		''() {
+			return this.parse('/j view-tournaments-all')
+		},
+		edit: 'add',
+		async add(target, room, user, connection, cmd) {
+			if (!toID(target).length) {
+				return this.parse(`/help smogtours`);
+			}
+			const targets = target.split('|');
+			const isEdit = cmd === 'edit';
+			const tourID = isEdit ? toID(targets.shift()) : null;
+			// {title}|{category}|{url}|{img}|{shortDesc}|{desc}
+			const [
+				title, rawSection, url, rawImg, rawShort, rawDesc,
+			] = Utils.splitFirst(targets.join('|'), '|', 5).map(f => f.trim());
+			const sectionID = toID(rawSection);
+			const section = tours[sectionID];
+			if (!section) {
+				return this.errorReply(`Invalid section ID: "${sectionID}"`);
+			}
+			checkCanEdit(user, this, sectionID);
+			// if we don't do this, it starts spitting out false every other time even if it's a valid link
+			//  since global regexes are stateful.
+			// this took me so much pain to debug.
+			// Yes, that is intended JS behavior. Yes, it fills me with unyielding rage.
+			// https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/RegExp/test
+			Chat.linkRegex.lastIndex = -1;
+			if (!Chat.linkRegex.test(url)) {
+				return this.popupReply(`Invalid info URL: "${url}"`);
+			}
+			let image;
+			if (rawImg) {
+				// see above
+				Chat.linkRegex.lastIndex = -1;
+				if (!Chat.linkRegex.test(rawImg)) {
```

Comment:
you might want to make a test function that sets `lastIndex` appropriately before using the regex

Reply:
ok